patcher.sh

```diff
@@ -101,16 +107,38 @@ patch_am2r ()
 		# Remove old lang folder
 		rm -R "$GAMEDIR"/lang
 
-		chmod +x "$GAMEDIR/runner"
-
-		if [ "$SYSTEMWIDE" = true ]; then
-			mkdir -p "$PREFIX/bin"
-			if [ "$APPIMAGE" = true ]; then
-				ln -sf "$GAMEDIR/AM2R.AppImage" "$PREFIX/bin/am2r"
-			else
-				ln -sf "$GAMEDIR/runner" "$PREFIX/bin/am2r"
-			fi
-		fi
+                # GameMaker games (like AMR2) link to OpenSSL 1.0.0, which is outdated and insecure.
+                # When attempting to link to newer versions an error is raised at runtime claiming it cannot find
+                # the outdated version of OpenSSL, even though it has been patched to link to the newer version.
+                # After replacing it with libcurl versioning is ignored, and the binary starts just fine.
+                echo "Replacing insecure OpenSSL dependency with libcurl..."
+                patchelf "$GAMEDIR/runner" \
+                        --replace-needed "libcrypto.so.1.0.0" "libcurl.so" \
+                        --replace-needed "libssl.so.1.0.0" "libcurl.so"
+
+                # An environment variable needs to be set on Mesa to avoid a race related to multithreaded shader compilation.
+                # To do this we move the original executable to a hidden file, and create a bash script with the needed variable in place of the original.
+                echo "Creating wrapper script to fix Mesa support..."
+                mv "$GAMEDIR/runner" "$GAMEDIR/.runner-unwrapped"
+                echo '
+                        #!/usr/bin/env bash
+                        # This environment variable fixes Mesa support. If another driver is used this should not do anything.
+                        # See https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/4181 for more information.
+                        radeonsi_sync_compile="true" exec "$(dirname "${BASH_SOURCE[0]}")/.runner-unwrapped" "$@"
+                ' | sed -e 's/^[[:space:]]*//' \
+                  | sed -r '/^\s*$/d' > "$GAMEDIR/runner"
+                # Note that the sed commands above remove leading whitespace and empty lines.
+
+                chmod +x "$GAMEDIR/runner" "$GAMEDIR/.runner-unwrapped"
+
+                if [ "$SYSTEMWIDE" = true ]; then
+                        mkdir -p "$PREFIX/bin"
+                        if [ "$APPIMAGE" = true ]; then
+                                ln -sf "$GAMEDIR/AM2R.AppImage" "$PREFIX/bin/am2r"
+                        else
+                                ln -sf "$GAMEDIR/runner" "$PREFIX/bin/am2r"
+                        fi
+                fi
```

Comment:
(Lowest line I can comment on, imagine this is the AppImage creation line)
In the current state, the AppImage still ships the outdated crypto/ssl libs. Do you want to remove those (in the patching process, the AM2RLauncher would break if you'd remove them from the repo)? AppImages aren't really sandboxed, so there may be security implications by them.

Reply:
> In the current state, the AppImage still ships the outdated crypto/ssl libs. Do you want to remove those (in the patching process, the AM2RLauncher would break if you'd remove them from the repo)? AppImages aren't really sandboxed, so there may be security implications by them.

I cannot test appimage creation myself, and I dont see where the insecure dependencies are included. I do see a bunch of `so`s being included in this repository, but ones compiled on NixOS would not work well in that environment. I would appreciate this if someone else could do that in a followup PR.

Reply:
Links to [libssl-100](https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux/blob/master/data/AM2R.AppDir/usr/lib32/libssl.so.1.0.0) and [libcrypto-100](https://github.com/AM2R-Community-Developers/AM2R-Autopatcher-Linux/blob/master/data/AM2R.AppDir/usr/lib32/libcrypto.so.1.0.0)
The whole AM2R.AppDir folder gets copied / used as a base for the AppImage.

> I would appreciate this if someone else could do that in a followup PR

Would also be fine for me.